### PR TITLE
New functionalities PR 1

### PR DESCRIPTION
**Pull request by ad-47**
Added new RPC for updateProjectName & updateProjectDescription
Rename message from LastModifiedExperimentRun to LastModifiedExperimentRun
Added two new RPC setProjectShortName && getProjectShortName
Improve getProjectSummary test case
Fix FindExperimentRun Test case
Changed date_time from string to uint64
Added 6 RPC for Single Tag Add/Delete support of Project & experiment & experimentRun

### CODE REVIEW

**Comment by ravishetye** [comment]
do you do anything different to build this?
I get error 
```[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.1:compile (default-compile) on project modeldb: Compilation failure: Compilation failure: 
[ERROR] /Users/ravishetye/work/open-source/modeldb-backend/target/generated-sources/protobuf/grpc-java/com/mitdbg/artifactstore/ArtifactStoreGrpc.java:[20,18] cannot find symbol
[ERROR]   symbol:   class Generated
[ERROR]   location: package javax.annotation
[ERROR] /Users/ravishetye/work/open-source/modeldb-backend/target/generated-sources/protobuf/grpc-java/com/mitdbg/modeldb/ProjectServiceGrpc.java:[20,18] cannot find symbol
[ERROR]   symbol:   class Generated
[ERROR]   location: package javax.annotation
[ERROR] /Users/ravishetye/work/open-source/modeldb-backend/target/generated-sources/protobuf/grpc-java/com/mitdbg/modeldb/JobServiceGrpc.java:[20,18] cannot find symbol
[ERROR]   symbol:   class Generated
[ERROR]   location: package javax.annotation
[ERROR] /Users/ravishetye/work/open-source/modeldb-backend/target/generated-sources/protobuf/grpc-java/com/mitdbg/modeldb/ExperimentRunServiceGrpc.java:[20,18] cannot find symbol
[ERROR]   symbol:   class Generated
[ERROR]   location: package javax.annotation
[ERROR] /Users/ravishetye/work/open-source/modeldb-backend/target/generated-sources/protobuf/grpc-java/com/mitdbg/modeldb/ExperimentServiceGrpc.java:[20,18] cannot find symbol
[ERROR]   symbol:   class Generated
[ERROR]   location: package javax.annotation

**Comment by ravishetye** [comment]
Did you run the tests for both rdb and mongo on this PR?

File: src/main/java/com/mitdbg/modeldb/ModelDBConstants.java [hunk 1]
```diff
@@ -20,6 +20,7 @@
   String KEY = "key";
   String VALUE = "value";
   String TAGS = "tags";
+  String SHORT_NAME = "short_name";
```
Comment (ravishetye) [inline src/main/java/com/mitdbg/modeldb/ModelDBConstants.java]: Can you verify if short name has test coverage ? It there is not can you please add so both in private and public repo.

File: src/main/java/com/mitdbg/modeldb/ModelDBUtils.java [hunk 2]
```diff
@@ -47,15 +57,52 @@ public static String getStringFromProtoObject(MessageOrBuilder object)
...
+      LOGGER.warning(e.getMessage());
+      Status status =
+          Status.newBuilder()
+              .setCode(Code.INTERNAL_VALUE)
+              .setMessage(e.getMessage())
+              .addDetails(Any.pack(CreateProject.Response.getDefaultInstance()))
+              .build();
+      throw StatusProto.toStatusRuntimeException(status);
+    }
+    if (projectShortName.length() > 42) {
```
Comment (ravishetye) [inline src/main/java/com/mitdbg/modeldb/ModelDBUtils.java]: extract as a ModelDBConstant

File: src/main/java/com/mitdbg/modeldb/ModelDBUtils.java [hunk 3]
```diff
@@ -47,15 +57,52 @@ public static String getStringFromProtoObject(MessageOrBuilder object)
...
+
+    // For specifying wrong message digest algorithms
+    catch (NoSuchAlgorithmException e) {
+      Status status =
+          Status.newBuilder().setCode(Code.INTERNAL_VALUE).setMessage(e.getMessage()).build();
+      throw StatusProto.toStatusRuntimeException(status);
+    }
+  }
+
+  public static String getURLEncoderString(String inputString) throws UnsupportedEncodingException {
```
Comment (ravishetye) [inline src/main/java/com/mitdbg/modeldb/ModelDBUtils.java]: can we make this private?

**Comment by ravishetye** [comment]
at few places in test Description is misspelt as Discription. Please fix it.

**Comment by ad-47** [comment]
> Did you run the tests for both rdb and mongo on this PR?

No, I just run maven clean and maven update from eclipse and run same commands which are we are using. nothing special to do.

File: src/main/java/com/mitdbg/modeldb/project/ProjectDAORdbImpl.java [hunk 4]
```diff
@@ -94,7 +90,7 @@ public Project insertProject(Project project) throws InvalidProtocolBufferExcept
 
   @Override
   @Transactional
-  public Project updateProject(String projectId, Project project)
+  public Project replaceProject(String projectId, Project project)
```
[the remark this answers is not in the document]
Comment (ad-47) [reply src/main/java/com/mitdbg/modeldb/project/ProjectDAORdbImpl.java]: this comment was added in the second PR.

**Comment by ad-47** [comment]
> do you do anything different to build this?
> I get error
> 
> ```
> [ERROR] /Users/ravishetye/work/open-source/modeldb-backend/target/generated-sources/protobuf/grpc-java/com/mitdbg/artifactstore/ArtifactStoreGrpc.java:[20,18] cannot find symbol
> [ERROR]   symbol:   class Generated
> [ERROR]   location: package javax.annotation
> [ERROR] /Users/ravishetye/work/open-source/modeldb-backend/target/generated-sources/protobuf/grpc-java/com/mitdbg/modeldb/ProjectServiceGrpc.java:[20,18] cannot find symbol
> [ERROR]   symbol:   class Generated
> [ERROR]   location: package javax.annotation
> [ERROR] /Users/ravishetye/work/open-source/modeldb-backend/target/generated-sources/protobuf/grpc-java/com/mitdbg/modeldb/JobServiceGrpc.java:[20,18] cannot find symbol
> [ERROR]   symbol:   class Generated
> [ERROR]   location: package javax.annotation
> [ERROR] /Users/ravishetye/work/open-source/modeldb-backend/target/generated-sources/protobuf/grpc-java/com/mitdbg/modeldb/ExperimentRunServiceGrpc.java:[20,18] cannot find symbol
> [ERROR]   symbol:   class Generated
> [ERROR]   location: package javax.annotation
> [ERROR] /Users/ravishetye/work/open-source/modeldb-backend/target/generated-sources/protobuf/grpc-java/com/mitdbg/modeldb/ExperimentServiceGrpc.java:[20,18] cannot find symbol
> [ERROR]   symbol:   class Generated
> [ERROR]   location: package javax.annotation
> ```

No, I just run maven clean and maven update from eclipse and run same commands which are we are using. nothing special to do.

File: src/main/java/com/mitdbg/modeldb/ModelDBConstants.java (hunk 1 above)
Comment (ad-47) [reply src/main/java/com/mitdbg/modeldb/ModelDBConstants.java]: We have already added test cases for short_name on both the branches.

File: src/main/java/com/mitdbg/modeldb/ModelDBUtils.java (hunk 2 above)
Comment (ad-47) [reply src/main/java/com/mitdbg/modeldb/ModelDBUtils.java]: Fixed

File: src/main/java/com/mitdbg/modeldb/ModelDBUtils.java (hunk 3 above)
Comment (ad-47) [reply src/main/java/com/mitdbg/modeldb/ModelDBUtils.java]: Fixed

File: src/main/java/com/mitdbg/modeldb/ModelDBConstants.java (hunk 1 above)
Comment (ad-47) [reply src/main/java/com/mitdbg/modeldb/ModelDBConstants.java]: Fixed

**Pull request merged**